app/src/main/java/fr/neamar/kiss/cache/MemoryCacheHelper.java

```diff
@@ -0,0 +1,165 @@
+                        .getDrawableIconForPackage(handle.componentName, handle.userHandle);
+                sAppIconCache.put(handle, drawable);
+            }
+        }
+        return drawable;
+    }
+
+    public static void cacheAppIconDrawable(@NonNull Context context, ComponentName className, UserHandle userHandle) {
+        AppIconHandle handle = new AppIconHandle(className, userHandle);
+        if (!sAppIconCache.containsKey(handle))
+            new AsyncAppIconLoad(context, handle, null).execute();
+    }
+
+    public static void trimMemory() {
+        synchronized (sAppIconCache) {
+            sAppIconCache.clear();
+        }
+        System.gc();
+    }
+
+    private static class AsyncAppIconLoad extends AsyncTask<Void, Void, Drawable> {
+        final WeakReference<Context> contextRef;
+        final AppIconHandle handle;
+        final OnDrawableReady callback;
+
+        public AsyncAppIconLoad(@NonNull Context context, AppIconHandle handle, @Nullable OnDrawableReady callback) {
+            this.contextRef = new WeakReference<>(context);
+            this.handle = handle;
+            this.callback = callback;
+        }
+
+        @Override
+        protected Drawable doInBackground(Void... voids) {
+            Context context = contextRef.get();
+            if (isCancelled() || context == null)
+                return null;
+            return getAppIconDrawable(context, handle.componentName, handle.userHandle);
+        }
+
+        @Override
+        protected void onPostExecute(Drawable drawable) {
+            super.onPostExecute(drawable);
+            if (isCancelled() || callback == null)
+                return;
+            callback.onDrawableReady(drawable);
+        }
+    }
+
+    private static class AppIconHandle implements Comparable<AppIconHandle> {
+        final ComponentName componentName;
+        final UserHandle userHandle;
+
+        AppIconHandle(ComponentName componentName, UserHandle userHandle) {
+            this.componentName = componentName;
+            this.userHandle = userHandle;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+
+            AppIconHandle handle = (AppIconHandle) o;
+
+            if (componentName != null ? !componentName.equals(handle.componentName) : handle.componentName != null)
+                return false;
+            return userHandle != null ? userHandle.equals(handle.userHandle) : handle.userHandle == null;
+        }
+
+        @Override
+        public int hashCode() {
+            int result = componentName != null ? componentName.hashCode() : 0;
+            result = 31 * result + (userHandle != null ? userHandle.hashCode() : 0);
+            return result;
+        }
+
+        @Override
+        public int compareTo(@NonNull AppIconHandle o) {
+            return 0;
```

Comment:
Why? This is missing some explanation for why all the `AppIconHandle`s are equal.

Reply:
Forgot about it :|